### ISSUE TITLE
podman doesn't handle filesystem-full situations gracefully

### ISSUE THREAD

**Issue by srcshelton** [containers/podman#10876]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

`podman system prune -f` should be usable in out-of-space conditions, however:

```
$ sudo podman system prune -f
Password:
Error: 235 errors occurred:
        * write /storage/containers/podman/storage/overlay-images/.tmp-images.json089675342: no space left on device
        * write /storage/containers/podman/storage/overlay-images/.tmp-images.json222771033: no space left on device
…
        * write /storage/containers/podman/storage/overlay-images/.tmp-images.json855797958: no space left on device
        * write /storage/containers/podman/storage/overlay-images/.tmp-images.json688457457: no space left on device
```

`podman` should make every precaution possible to be able to clean up after itself even in filesystem-full situations.

In this case, simply removing two empty directories was enough to allow a second invocation of `podman system prune` to do the right thing.

**Output of `podman version`:**

```
Version:      3.2.2
API Version:  3.2.2
Go Version:   go1.16.5
Git Commit:   d577c44e359f9f8284b38cf984f939b3020badc3
Built:        Mon Jul  5 00:13:23 2021
OS/Arch:      linux/amd64
```

**Output of `podman info --debug`:**

```
host:
  arch: amd64
  buildahVersion: 1.21.0
  cgroupControllers:
  - cpuset
  - cpu
  - io
  - memory
  - hugetlb
  - pids
  - rdma
  cgroupManager: cgroupfs
  cgroupVersion: v2
  conmon:
    package: app-emulation/conmon-2.0.29
    path: /usr/bin/conmon
    version: 'conmon version 2.0.29, commit: 7e6de6678f6ed8a18661e1d5721b81ccee293b9b'
  cpus: 4
  distribution:
    distribution: gentoo
    version: unknown
  eventLogger: file
  hostname: apu2e
  idMappings:
    gidmap: null
    uidmap: null
  kernel: 5.12.0-gentoo
  linkmode: dynamic
  memFree: 1828438016
  memTotal: 4116377600
  ociRuntime:
    name: crun
    package: app-emulation/crun-0.20.1
    path: /usr/bin/crun
    version: |-
      crun version 0.20.1
      commit: 38271d1c8d9641a2cdc70acfa3dcb6996d124b3d
      spec: 1.0.0
      +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +YAJL
  os: linux
  remoteSocket:
    path: /run/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_AUDIT_WRITE,CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_MKNOD,CAP_NET_BIND_SERVICE,CAP_NET_RAW,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: false
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: false
  serviceIsRemote: false
  slirp4netns:
    executable: ""
    package: ""
    version: ""
  swapFree: 2434490368
  swapTotal: 2684350464
  uptime: 172h 25m 4.41s (Approximately 7.17 days)
registries:
  localhost:5000:
    Blocked: false
    Insecure: true
    Location: localhost:5000
    MirrorByDigestOnly: false
    Mirrors: []
    Prefix: localhost:5000
  search:
  - docker.io
  - docker.pkg.github.com
  - quay.io
  - public.ecr.aws
store:
  configFile: /etc/containers/storage.conf
  containerStore:
    number: 2
    paused: 0
    running: 2
    stopped: 0
  graphDriverName: overlay
  graphOptions:
    overlay.ignore_chown_errors: "false"
  graphRoot: /storage/containers/podman/storage
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "true"
  imageStore:
    number: 17
  runRoot: /storage/containers/podman/run
  volumePath: /storage/containers/podman/volumes
version:
  APIVersion: 3.2.2
  Built: 1625440403
  BuiltTime: Mon Jul  5 00:13:23 2021
  GitCommit: d577c44e359f9f8284b38cf984f939b3020badc3
  GoVersion: go1.16.5
  OsArch: linux/amd64
  Version: 3.2.2

```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide? (https://github.com/containers/podman/blob/master/troubleshooting.md)**

Yes

**Comment by rhatdan** [comment]
I am not sure how podman figures out what the right thing to do would be here?

**Comment by baude** [comment]
is the problem here that c/storage needs a file lock but the filesystem is 100% full?

**Comment by srcshelton** [comment]
> I am not sure how podman figures out what the right thing to do would be here?

I guess it's a case of - if we know we're wanting to remove objects - trying to free up space before writing anything?

For a delete operation, the process could even try to take a lock but make any -ENOSPC error non-fatal, or have a `--no-lock` option to perform a slightly unsafe deletion if necessary?

**Comment by baude** [comment]
@srcshelton we talked about this issue today as a team.  we have some ideas on how to possibly handle this.  it would be helpful to us if you could describe how you get into this situation.  thanks in advance!

**Comment by srcshelton** [comment]
> @srcshelton we talked about this issue today as a team. we have some ideas on how to possibly handle this. it would be helpful to us if you could describe how you get into this situation. thanks in advance!

Nothing _particularly_ out of the ordinary, I don't think - I've been working on and off for just over a year now to migrate the services I run on my machines at home to running in containers.

The main infrastructure server has long been done, and I've turned my attention to some more modest devices: an 8GB 4x2.1GHz Raspberry Pi 4, and a 4GB 4x1GHz AMD Jaguar system running as a router.  Both are equipped with SSDs, however not especially large ones.

In addition (and this may be a bug - I can add further details in another ticket if desired) podman from 2.0 and before right up to the latest release have always left a significant number of intermediate(?)  '`<none>`' images visible if you do a `podman image ls`, and an even greater number of images (or image fragments?) visible from `buildah images` - and performing a `podman image prune` can regularly result in many tens if not hundreds of GB of data in these artefacts being freed.

In this case, I'd accidentally omitted to run a `podman image prune` or `podman system prune` operation before starting an update-build on the AMD-based system, and it ran out of disk-space on the volume (`/storage`) containing the container graph root (my `storage.conf` on this host contains:

```
[storage]
driver = "overlay"
runroot = "/storage/containers/podman/run"
graphroot = "/storage/containers/podman/storage"
rootless_storage_path = "/storage/containers/podman/user/$USER"
```

… although all of my current usage is rootful).

Which I was able to recover without too much drama when I found a couple of empty nested directories beneath `/storage/containers/podman/tmp` - but, had this not been the case, I'd have been having to guess and manually (re)move something podman related in order for the prune operation to work with a full filesystem (as the only data which exists on this partition is all beneath `/storage/containers/podman`).

If there were a way that `podman` could detect the case where there is insufficient disk-space to even take a lock and either use an alternate fall-back lock directory or possibly perform a one-time lockless deletion in order to free the space to write further locks, then this would make the situation much more recoverable from a user point of view.

Out of interest, the project I've been working on for this time (which is still very WIP!) is at https://github.com/srcshelton/docker-gentoo-build, and the `gentoo-init.docker` script (… whose name was chosen before I migrated to `podman`) at least should be able to run from most systems capable of running `podman`!

**Comment by vrothberg** [comment]
@baude, do you recall the team's ideas?

**Comment by giuseppe** [comment]
I don't think Podman needs to deal with this issue.  If you run out of space, then you need to clean up something to make it possible for Podman to work.

If you know you are going to end up without any space left, one possible solution is `touch /var/lib/containers/storage/{remove_me_if_needed,remove_me_if_needed2}` that you can delete when needed.

I am closing the issue, but please reopen if you disagree and think Podman needs to care of it

**Comment by vrothberg** [comment]
I am torn since I sympathize with the expectation that `podman system prune -f` would be able to free up space.

> write /storage/containers/podman/storage/overlay-images/.tmp-images.json089675342

I am not sure _why_ writing a temporary images.json would trigger that error though.  Shouldn't the images have been free'd prior to writing that?  Could it be a kernel fart that the write is scheduled prior to deleting the image(s)?
